src/transformers/utils/import_utils.py

```diff
@@ -1538,6 +1598,20 @@ def __getattr__(self, name: str) -> Any:
             return self._objects[name]
         if name in self._modules:
             value = self._get_module(name)
+        elif name in self._object_missing_backend.keys():
+            missing_backends = self._object_missing_backend[name]
+
+            class Placeholder(metaclass=DummyObject):
+                _backends = missing_backends
+
+                def __init__(self, *args, **kwargs):
+                    requires_backends(self, missing_backends)
```

Comment:
I find the logic a bit complex here:
- we have a method `requires_backends` that takes an input an obj and a list of backends
    - the `obj` is forwarded only to retrieve the name of the class/of the object ([L1473](https://github.com/huggingface/transformers/blob/d805428270e4c6ca4dc6b240567e87c3916f957c/src/transformers/utils/import_utils.py#L1473)) 
- we have a `DummyObject` metaclass to define `__getattributes__` => fails when accessing any public property/method
- we have a `Placeholder` object using the metaclass + defining a dummy `__init__` method that also requires backends.
    - `Placeholder.__name__` and `Placeholder.__module__` are then set afterward.

I feel that this is a bit clunky. The ImportError logic is set twice (once in metaclass, one in class). Some class attribute are set in class definition (`_backends`) while others are set afterwards (`__name__`/`__module__`). `_backends` is set only so that the metaclass sees it, etc. 


Something you can do to dynamically generate classes is to still use a metaclass but use is as a callable. It's quite low-level Python but what you are doing here is low-level anyway. Something like this should work:

```py
class PlaceholderFactory(type):

    # 1. check that `_backends` is provided
    def __new__(cls, name: str, bases: Any, namespace: Any) -> Any:
        if "_backends" not in namespace:
            raise RuntimeError("`_backends` must be provided when generating a class with `PlaceholderFactory`.")
        return super().__new__(cls, name, bases, namespace)

    # 2. forbid to call `__init__`
    def __call__(cls, *args, **kwargs) -> Any:
        cls._requires_backends()

    # 3. forbid any class attributes
    def __getattribute__(cls, name: str) -> Any:
        if "_" in name:
            return super().__getattribute__(name)
        cls._requires_backends()

    # 4.  alias for `requires_backends` but all the logic can also be moved inside it
    def _requires_backends(cls):
        requires_backends(cls.__name__, cls._backends)
```

Once you have this, you can dynamically create the placeholder class like this:

```py
value = PlaceholderFactory(name, (), {"__module__": self.__spec__, "_backends": missing_backends})
```

IMO it makes it clearer which part is handling what.

Reply:
I think this is a very valuable comment with which I entirely agree. I would be happy for us to get to that in a follow-up PR if that works with you.

Reply:
of course!